app/src/backlight.c

```diff
@@ -0,0 +1,237 @@
+    led_bl = device_get_binding(BACKLIGHT_LABEL);
+    if (led_bl) {
+        LOG_INF("Found LED backlight device %s", BACKLIGHT_LABEL);
+    } else {
+        LOG_ERR("LED backlight device %s not found", BACKLIGHT_LABEL);
+        return -EINVAL;
+    }
+
+    state = (struct backlight_state){
+        brightness : CONFIG_ZMK_BACKLIGHT_LEVEL_START - 1,
+        on : IS_ENABLED(ZMK_BACKLIGHT_ON_START)
+    };
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+    settings_subsys_init();
+
+    int err = settings_register(&bl_conf);
+    if (err) {
+        LOG_ERR("Failed to register the backlight settings handler (err %d)", err);
+        return err;
+    }
+
+    k_delayed_work_init(&backlight_save_work, zmk_backlight_save_state_work);
+
+    settings_load_subtree("backlight");
+#endif
+
+    return zmk_backlight_update();
+}
+
+static int zmk_backlight_set_on(bool on) {
+    if (!led_bl) {
+        return -ENODEV;
+    }
+
+    if (on && state.brightness == 0) {
+        state.brightness++;
+    }
+    state.on = on;
+
+    int rc = zmk_backlight_update();
+    if (rc != 0) {
+        return rc;
+    }
+
+    return zmk_backlight_save_state();
+}
+
+static int zmk_backlight_change_brt(int direction) {
+    if (!led_bl) {
+        return -ENODEV;
+    }
+
+    int b = state.brightness + direction;
+    if (b < 0) {
+        b = 0;
+    } else if (b >= BACKLIGHT_LEVELS) {
+        b = BACKLIGHT_LEVELS - 1;
+    }
+
+    // turn on backlight when change brightness
+    state.on = (b > 0);
+    state.brightness = b;
+
+    int rc = zmk_backlight_update();
+    if (rc != 0) {
+        return rc;
+    }
+
+    return zmk_backlight_save_state();
+}
+
+int zmk_backlight_toggle() { return zmk_backlight_set_on(!state.on); }
+
+int zmk_backlight_on() { return zmk_backlight_set_on(true); }
+
+int zmk_backlight_off() { return zmk_backlight_set_on(false); }
+
+int zmk_backlight_inc() { return zmk_backlight_change_brt(1); }
+
+int zmk_backlight_dec() { return zmk_backlight_change_brt(-1); }
+
+#if IS_ENABLED(ZMK_BACKLIGHT_SLEEP) || IS_ENABLED(ZMK_BACKLIGHT_USB)
+static int backlight_event_handler(const zmk_event_t *eh) {
+    if (!led_bl) {
+        return -ENODEV;
+    }
+
+#if IS_ENABLED(ZMK_BACKLIGHT_SLEEP)
+    struct zmk_activity_state_changed *ev_act = as_zmk_activity_state_changed(eh);
+    if (ev_act != NULL) {
+        switch (ev_act->state) {
+        case ZMK_ACTIVITY_ACTIVE:
+            zmk_backlight_set_on();
+            break;
+        case ZMK_ACTIVITY_IDLE:
+        case ZMK_ACTIVITY_SLEEP:
+            zmk_backlight_set_off();
+            break;
+        default:
+            LOG_WRN("Unhandled activity state: %d", ev_act->state);
+            return -EINVAL;
+        }
+
+        return 0;
+    }
+#endif // IS_ENABLED(ZMK_BACKLIGHT_SLEEP)
+
+#if IS_ENABLED(ZMK_BACKLIGHT_USB)
+    struct zmk_usb_conn_state_changed *ev_usb = as_zmk_usb_conn_state_changed(eh);
+    if (ev_usb != NULL) {
+        switch (ev_usb->state) {
```

Comment:
Has this been tested? It doesn't look like this will compile. `zmk_usb_conn_state_changed` has a member named `conn_state`, not `state`.

Also, you could maybe simplify this whole thing to just

```c
if (zmk_usb_is_powered()) {
  zmk_backlight_set_on();
} else {
  zmk_backlight_set_off();
}
```
since the value of `ev_usb->conn_state` comes from the state variable that `zmk_usb_is_powered()` uses.

Since `ev_usb` wouldn't be used at all except to check the event type in that case, you could also simplify the `if` check to 

```c
if (as_zmk_usb_conn_state_changed(eh))
```
which admittedly is a bit awkward to read. You could instead set up a separate `ZMK_LISTENER()` for each event so you don't have to check the event type at all.